### DISCUSSION TITLE
Allow system admin to see what is collected

### DISCUSSION THREAD

**Discussion by djoreilly** [Velocidex/velociraptor#3644] [Ideas]
Sometimes system admins need to know what is being collected by Velociraptor on their endpoints.  They would like to be able to verify that no sensitive data is being sent to the server. Maybe a client side option that would write the VQL responses to JSON files.

**Comment by scudette** [comment]
You can set up a server monitoring artifact to view all completed collections - you can see an example of such a thing here https://docs.velociraptor.app/artifact_references/pages/elastic.flows.upload/ which actually uploads the data. 

You can also see all the initiated collections and hunt in the audit log (and forward it to another system or maybe slack or discord in real time). All audit events are sent on this artifact https://docs.velociraptor.app/artifact_references/pages/server.audit.logs/ so you can monitor it with `SELECT * FROM watch_monitoring(artifact="Server.Audit.Logs")`

**Comment by djoreilly** [comment]
So I mean the OS system administrators of the endpoint machines running Velociraptor clients - not the Velociraptor administrators. Even if forwarding to Elastic was set up, there would need some way to ensure they only have visibility to the data from their endpoints.

**Comment by djoreilly** [comment]
For some more context here is an example of a concern sysadmins have. There may be something running on their endpoint that shells out and puts sensitive information on the command line and `Linux.Events.ProcessExecutions` will send it to the Velociraptor server. A feature like this allows them to monitor what is being gathered for a while until they are satifised that nothing sensitive is leaking out.

I have done a quick implementation https://github.com/djoreilly/velociraptor/commit/2ede58e031bb6e5294ebeb93ee19216141a33c42

**Reply by scudette** [reply]
Thanks for clarifying. Normally leaving unencrypted information on the endpoint is not considered best practice. This is why typically the client does not log anything.


Although in this case it's similar to just logging process execution locally. We can store local encrypted logs now which might make it better.

Is this really a discussion about access control and visibility? The data is available on the server anyway so is the use case to provide real only access to view the data?

**Reply by djoreilly** [reply]
> Thanks for clarifying. Normally leaving unencrypted information on the endpoint is not considered best practice. This is why typically the client does not log anything.
>> It could be an option disabled by default and come with a security warning. It would be up to the system admin if they want to go ahead with the risk. So it would be like `--verbose` client logging - only to be used judiciously.
> 
> Although in this case it's similar to just logging process execution locally. We can store local encrypted logs now which might make it better.
>> The process executions is just one example. Some sysadmins have general concerns and would like to see everything. About the encryption, how would the key be stored locally?
> 
> Is this really a discussion about access control and visibility? The data is available on the server anyway so is the use case to provide real only access to view the data?
>> Sure the data is available on the server, but the security team may not want to give all these system admins access to the Velociraptor GUI. Even with a read-only role they would be able to see information from endpoints that they don't administer.

**Reply by scudette** [reply]
What exactly do you want to log ? the requests or the results? 

The requests can be logged in the clear using the regular logging mechanism (by enabling client side logging). You can also access the client side logs (on the client itself) using the logging() plugin https://docs.velociraptor.app/vql_reference/misc/logging/ so you could do anything with the logs including write them locally. All you would need to do is write a client monitoring artifact that collects client side logs and writes them somewhere- for example https://docs.velociraptor.app/artifact_references/pages/generic.client.locallogs/

The encrypted local logs are designed to be only decryptable on the server so they are secure to be stored locally on the endpoint. If you want the machine owner to be able to decrypt the logs you can write the results into a PKI encrypted container (e.g. PGP keys or X509) so they can decrypt it off the system. See the collect() plugin https://docs.velociraptor.app/vql_reference/plugin/collect/ (this is basically how the offline collector works).

I dont think this will make it very easy to use though. Practically how would the system owner get the logs off the system? 

Would it just make more sense to send the system owner an email summarising all actions on their endpoint from the server? This can be sent via slack, discord or email and can include information such as who actually scheduled the collection and why. It seems like a better auditing solution than writing the information on the endpoint itself.

**Reply by djoreilly** [reply]
> What exactly do you want to log ? the requests or the results?

The results. 

> Would it just make more sense to send the system owner an email summarising all actions on their endpoint from the server? This can be sent via slack, discord or email and can include information such as who actually scheduled the collection and why. It seems like a better auditing solution than writing the information on the endpoint itself.

I think they just want to see the raw results. Not just results from collections, but results from client monitoring artifacts too.

How about something like this - an option to log the vql results at debug level?
https://github.com/djoreilly/velociraptor/commit/87f5a80eb706a6067060ac14f9f25763af8fbd46

**Reply by scudette** [reply]
There is no point as your pr is basically the same as writing a vql client side monitoring to write local logs anyway. 

Check out the local logs artifact I pointed to above and see how to modify it to do the same thing

**Reply by djoreilly** [reply]
I tried this but the results of the VQL queries are not logged - it's the same as running the client with `--verbose`. Am I missing something? Thanks for your help.

```
name: Custom.Generic.Client.LocalLogs2

description: |
  Write client logs to a local file

type: CLIENT_EVENT

parameters:
- name: LocalFilename
  default: "/tmp/test-local-logs.log"
  description: The local filename that will be written.
- name: AlsoForward
  type: bool
  description: |
    By default we do not forward any of the logs to the server but
    this allows logs to be forwarded as well as written locally.
- name: Component
  default: generic
  description: The log component to forward (default "generic")
  type: choices
  choices:
    - generic
    - client
    - frontend
    - gui
    - api

sources:
- query: |
     LET _ <= log(message="Writing local log to " + expand(path=LocalFilename))

     SELECT * FROM write_jsonl(
       filename=expand(path=LocalFilename),
       buffer_size=10000,
       query={
         SELECT timestamp(epoch=now()) AS Timestamp, *
         FROM logging(component=Component)
       })
     WHERE AlsoForward
```

**Reply by scudette** [reply]
Yes this is correct - you are writing the output of the logs which are the same as shown with the --verbose flag. 

It is not reasonable IMHO to log the actual output from the queries (some queries generate gb of results). The logs show what kind of data was collected though. What is the purpose of the logs? to see if the collection was reasonable or misused?

**Reply by djoreilly** [reply]
> It is not reasonable IMHO to log the actual output from the queries (some queries generate gb of results). The logs show what kind of data was collected though. What is the purpose of the logs? to see if the collection was reasonable or misused?

We have a policy that all Linux servers on the company network must have the Velociraptor client installed. This makes some system owners a bit anxious and they wish to know exactly what is collected. Sure we can point them to the source code for all the artifacts we use, but most people don't know VQL and would have difficultly understanding them. If there was an option like this, then they could enable it for as long as they like and just look at the raw data that is being sent to the server. Of course this option would not be set in the default client config pushed out to all clients. I have updated the diff so the option comes with a warning https://github.com/djoreilly/velociraptor/commit/2f1751c535012e26f20122d3faf590eaa85ad58f

**Reply by scudette** [reply]
The problem with the PR is that it is sending potentially very large results into the logs. Velociraptor keeps the last 1000 log messages in memory so they can be accessed remotely during profile collection or debugging. So this data will stay around for quite a while and may cause high memory usage.  Typically logs should not be handling very large messages.

**Reply by djoreilly** [reply]
Ah okay, I didn't know the last 1000 log messages where kept in memory. I'll get back to the people asking for the feature and explain the problems. Thanks again.

**Reply by scudette** [reply]
You can see where that is done here
https://github.com/Velocidex/velociraptor/blob/f753569928b4df7d294d0f4a5785793b2bd9b9dd/logging/logging.go#L542

You can just write the data out yourself (i.e. not through the log)
